waffle-doppelganger/src/index.ts

```diff
@@ -0,0 +1,43 @@
+import {utils, Contract, Wallet, ContractFactory} from 'ethers';
+import {FunctionDescription} from 'ethers/utils/interface';
+import {EventFragment, FunctionFragment, ParamType} from 'ethers/utils/abi-coder';
+import DoppelgangerContract from '../build/Doppelganger.json';
+
+type ABI = Array<EventFragment | FunctionFragment | ParamType> | string;
+
+async function deployDoppelganger(wallet: Wallet) {
+  const factory = new ContractFactory(DoppelgangerContract.abi, DoppelgangerContract.bytecode, wallet);
+  return factory.deploy();
+}
+
+function stub(mockContract: Contract, encoder: utils.AbiCoder, func: FunctionDescription) {
+  return {
+    returns: async (...args: any) => {
+      const encoded = encoder.encode(func.outputs, args);
+      await mockContract.mockReturns(func.sighash, encoded);
+    }
+  };
+}
+
+export type Doppelganger = {
```

Comment:
What do you think about
```ts
interface Doppelganager extends Contract {
  mock: {
    [key: string]: {
      returns: (value: unknown) => Promise<void>
      // potentially in the future
      reverts: (reason: string) => Promise<void>
    }
  }
}
```

Reply:
To be fair, I've considered this and don't have a strong opinion about any of the options - so it's your call.

Reply:
I think I'd go for this version. It is more intuitive to have the `doppelganger` or `mockContract` function return a contract